common/js/injected_script.js

```diff
@@ -143,6 +155,43 @@ class XPathQueryEngine {
   }
 }
 
+// convertToDocument will convert a DocumentFragment into a Document. It does
+// this by creating a new Document and copying the elements from the
+// DocumentFragment to the Document.
+function convertToDocument(fragment) {
+  var newDoc = document.implementation.createHTMLDocument("Temporary Document");
+
+  copyNodesToDocument(fragment, newDoc.body);
+
+  return newDoc;
+}
+
+// copyNodesToDocument manually copies nodes to a new document, excluding
+// ShadowRoot nodes -- ShadowRoot are not cloneable so we need to manually
+// clone them one element at a time.
+function copyNodesToDocument(sourceNode, targetNode) {
```

Comment:
Should we add a "too deep" error here? Think about what would happen if the element is super hierarchical that might clog our system. That might lead to instability issues.

Reply:
Yeah, that's a good point. The issue is how far deep do we allow this to go? This is where my lack of frontend knowledge hinders me in working out a good valid depth. Maybe it's ok to allow it to go until an OOM error occurs, and work backwards from there? Any ideas?

Reply:
I'm going to merge this in, but happy to discuss and work on this if you think it's an issue. My guess is that we might not need to worry for majority of use cases. I would imagine that Chrome would detect layouts which could go too deep before the browser module gets a chance.

Reply:
Sure, let's wait for it to happen first.

Reply:
@ankur22 

Sure, let's wait for it to happen first. I hope we can detect this is the cause if it ever happens 🤞